File: control-plane/internal/domain/entity/l4.go
package entity

// Tuân thủ Flat Entity: không chứa json tags.

// L4ACLRule biểu diễn một quy tắc kiểm soát truy cập IP/CIDR ở tầng L4.
type L4ACLRule struct {
	CIDR        string
	Action      string // "allow" | "deny"
	Description string
}

// L4ServiceItem là flat projection entity đại diện cho một L4 Service (Port Listener & ACL).
type L4ServiceItem struct {
	ID                  string
	Name                string
	Protocol            string
	ListenPort          int
	ForwardTargetType   string // "upstream" | "endpoint"
	UpstreamName        string
	DirectEndpoint      string
	ACLRulesJSON        string
	ProxyTimeout        string
	ProxyConnectTimeout string
	Enabled             bool
	Description         string
	CreatedAt           string
	UpdatedAt           string
}

// CreateL4ServiceCommand chứa dữ liệu tạo mới L4 Service.
type CreateL4ServiceCommand struct {
	ID                  string
	Name                string
	Protocol            string
	ListenPort          int
	ForwardTargetType   string // "upstream" | "endpoint"
	UpstreamName        string
	DirectEndpoint      string
	ACLRulesJSON        string
	ProxyTimeout        string
	ProxyConnectTimeout string
	Enabled             bool
	Description         string
}

// UpdateL4ServiceCommand chứa dữ liệu cập nhật L4 Service.
type UpdateL4ServiceCommand struct {
	ID                  string
	Name                string
	Protocol            string
	ListenPort          int
	ForwardTargetType   string // "upstream" | "endpoint"
	UpstreamName        string
	DirectEndpoint      string
	ACLRulesJSON        string
	ProxyTimeout        string
	ProxyConnectTimeout string
	Enabled             bool
	Description         string
}

// ListL4ServicesQuery chứa điều kiện tìm kiếm L4 Services.
type ListL4ServicesQuery struct {
	Search   string
	Protocol string
	Limit    int
	Offset   int
}

File: ui/src/lib/api/l4.ts
import { api } from '../fetcher';

export interface L4ACLRule {
  cidr: string;
  action: 'allow' | 'deny';
  priority: number;
  description?: string;
}

export interface L4ServiceItem {
  id: string;
  name: string;
  protocol: string; // 'tcp' | 'udp'
  listen_port: number;
  forward_target_type: 'upstream' | 'endpoint';
  upstream_name: string;
  direct_endpoint: string;
  acl_rules_json: string;
  proxy_timeout: string;
  proxy_connect_timeout: string;
  enabled: boolean;
  description: string;
  created_at: string;
  updated_at: string;
}

export interface ListL4ServicesResponse {
  items: L4ServiceItem[];
  total: number;
}

export interface CreateL4ServicePayload {
  name: string;
  protocol: string;
  listen_port: number;
  forward_target_type?: 'upstream' | 'endpoint';
  upstream_name?: string;
  direct_endpoint?: string;
  acl_rules_json: string;
  proxy_timeout?: string;
  proxy_connect_timeout?: string;
  enabled?: boolean;
  description?: string;
}

export interface UpdateL4ServicePayload {
  name: string;
  protocol: string;
  listen_port: number;
  forward_target_type?: 'upstream' | 'endpoint';
  upstream_name?: string;
  direct_endpoint?: string;
  acl_rules_json: string;
  proxy_timeout?: string;
  proxy_connect_timeout?: string;
  enabled?: boolean;
  description?: string;
}

export const l4Api = {
  listServices: (params?: Record<string, string | number | boolean | null | undefined>) =>
    api.get<ListL4ServicesResponse>('/api/v1/l4/services', params),
  getService: (id: string) => api.get<L4ServiceItem>(`/api/v1/l4/services/${id}`),
  createService: (payload: CreateL4ServicePayload) =>
    api.post<L4ServiceItem>('/api/v1/l4/services', payload),
  updateService: (id: string, payload: UpdateL4ServicePayload) =>
    api.put<L4ServiceItem>(`/api/v1/l4/services/${id}`, payload),
  toggleService: (id: string, enabled: boolean) =>
    api.put<L4ServiceItem>(`/api/v1/l4/services/${id}/status`, { enabled }),
  deleteService: (id: string) =>
    api.delete<{ message: string }>(`/api/v1/l4/services/${id}`),
};
